Add SecureBoot support for arbitrary operating systems to "Grub2 UEFI" PXE loaders (#2145)
theforeman/foreman-documentation: guides/common/modules/con_using-pxe-to-provision-hosts.adoc

endif::[]

In {Project} provisioning, the PXE loader option defines the DHCP `filename` option to use during provisioning.
For BIOS systems, select the *PXELinux BIOS* option to enable a provisioned node to download the `pxelinux.0` file over TFTP.

> For UEFI systems, select the *Grub2 UEFI* option to enable a TFTP client to download `grubx64.efi` file, or select the *Grub2 UEFI HTTP* option to enable an UEFI HTTP client to download `grubx64.efi` with the HTTP Boot feature.

Reviewer: This is outside the scope of this PR. Just FYI. 

GRUB is an acronym. The UI labels should be "GRUB2 ..."

In {Project} provisioning, the PXE loader option defines the DHCP `filename` option to use during provisioning.
For BIOS systems, use the *PXELinux BIOS* option to enable a provisioned node to download the `pxelinux.0` file over TFTP.

> For UEFI systems, use the *Grub2 UEFI* option to enable a TFTP client to download `grub2/grubx64.efi` file, or use the *Grub2 UEFI HTTP* option to enable an UEFI HTTP client to download `grubx64.efi` from {SmartProxy} with the HTTP Boot feature.

Reviewer: I am not sure what "use the <option>" means. Is the user responding to CLI prompts? Or is this a UI? 

It might be better to use "Select", depending on the context.

> For UEFI systems, use the *Grub2 UEFI* option to enable a TFTP client to download `grub2/grubx64.efi` file, or use the *Grub2 UEFI HTTP* option to enable an UEFI HTTP client to download `grubx64.efi` from {SmartProxy} with the HTTP Boot feature.

Reviewer 2: `PXEGrub2` is the internal identifier of the PXE laoder. But if you have a look at the Web UI e.g. in `Hosts` > `Create Hosts` > `Operating System` > `PXE loader` in this list a "display name" is used. And that is `Grub2 UEFI` resp. `Grub2 UEFI HTTP`. In the documentation we should use the names visible to users in the Web UI.

> For UEFI systems, use the *Grub2 UEFI* option to enable a TFTP client to download `grub2/grubx64.efi` file, or use the *Grub2 UEFI HTTP* option to enable an UEFI HTTP client to download `grubx64.efi` from {SmartProxy} with the HTTP Boot feature.

Reviewer 2: Regarding "use the" versus "select": I would prefer "select", as it's in the Web UI and you select the PXE loader from a dropdown selection.

> For UEFI systems, use the *Grub2 UEFI* option to enable a TFTP client to download `grub2/grubx64.efi` file, or use the *Grub2 UEFI HTTP* option to enable an UEFI HTTP client to download `grubx64.efi` from {SmartProxy} with the HTTP Boot feature.

Reviewer 2: "use the" => "select the" :heavy_check_mark:

ifndef::satellite[]
Use SecureBoot options to enable a client to download the `shim.efi` bootstrap bootloader that then loads the signed `grubx64.efi`.
endif::[]

ifndef::orcharhino,satellite[]

> Other PXE loaders like PXELinux UEFI, Grub2 ELF or iPXE Chain, require additional configuration. These workflows are not documented at the moment.

Reviewer 3: ```suggestion
Other PXE loaders like PXELinux UEFI, Grub2 ELF or iPXE Chain, require additional configuration.
These workflows are not documented at the moment.
```

Unrelated to your PR.

endif::[]

For BIOS provisioning, you must associate a PXELinux template with the operating system.
For UEFI provisioning, you must associate a PXEGrub2 template with the operating system.
If you associate both PXELinux and PXEGrub2 templates, {Project} can deploy configuration files for both on a TFTP server, so that you can switch between PXE loaders easily.
ifndef::satellite[]

.Provisioning of Secure Boot enabled hosts
By default, you can provision Secure Boot enabled hosts that match the operating system of your {ProjectServer}.

> You have to use the SecureBoot variants of the *Grub2 UEFI* and *Grub2 UEFI HTTP* PXE loaders to enable a client to download the `shim.efi` bootstrap bootloader that then loads the signed `grubx64.efi`.

Reviewer 3: ```suggestion
You have to use the SecureBoot variants of the *Grub2 UEFI* and *Grub2 UEFI HTTP* PXE loaders to enable a host to download the `shim.efi` bootstrap bootloader that then loads the signed `grubx64.efi`.
```

> You have to use the SecureBoot variants of the *Grub2 UEFI* and *Grub2 UEFI HTTP* PXE loaders to enable a client to download the `shim.efi` bootstrap bootloader that then loads the signed `grubx64.efi`.

Reviewer: Global recommendation: Change "You have" to "You must" to align with the IBM Style Guide. 

"You have" can create problems with localization and translations into other languages because "have" has many different meanings.

Reviewer 2: Fixed

To provision other operating systems than the operating system of your {ProjectServer} on Secure Boot enabled hosts, you have to provide the `shim` and `GRUB 2` binaries for each operating system you want to provision.

> For further information about the required configuartion steps, see:

Reviewer: Typo: "configuartion"

Reviewer 2: Fixed

* xref:Configuring-{project-context}-Server-to-Provision-almalinux-on-Secure-Boot-Enabled-Hosts_{context}[]
* xref:Configuring-{project-context}-Server-to-Provision-debian-on-Secure-Boot-Enabled-Hosts_{context}[]
* xref:Configuring-{project-context}-Server-to-Provision-rocky-on-Secure-Boot-Enabled-Hosts_{context}[]
* xref:Configuring-{project-context}-Server-to-Provision-ubuntu-on-Secure-Boot-Enabled-Hosts_{context}[]
endif::[]

> Afterwards use the *Grub2 UEFI SecureBoot (target OS)* option to provision these hosts.

Reviewer 3: ```suggestion
After configuring your {ProjectServer}, you can then use the *Grub2 UEFI SecureBoot (target OS)* option to provision Secure Boot enabled hosts.
```

For BIOS provisioning, you must associate a PXELinux template with the operating system.
For UEFI provisioning, you must associate a PXEGrub2 template with the operating system.
If you associate both PXELinux and PXEGrub2 templates, {Project} can deploy configuration files for both on a TFTP server, so that you can switch between PXE loaders easily.

.SecureBoot Support

SecureBoot expects to follow a chain of trust from the start of the host to the loading of Linux kernel modules.
The very first shim that is loaded basically determines which distribution is allowed to be booted or kexec’ed until next reboot.

> The following assumes host systems with enabled SecureBoot (User Mode) and default platform certificates.

Author: provides

If you are using custom keys, you have to sign the corresponding components accordingly.

> To make this POC working you need to provide all shim and GRUB2 binaries manually under `/usr/local/share/bootloader-universe/<os>/`.

Reviewer 2: ```suggestion
To make provisioning of UEFI systems with enabled SecureBoot work, you need to provide all shim and GRUB2 binaries manually under `/usr/local/share/bootloader-universe/<os>/`.
```

> To make this POC working you need to provide all shim and GRUB2 binaries manually under `/usr/local/share/bootloader-universe/<os>/`.

Reviewer 3: ```suggestion
* All shim and GRUB2 binaries are present in `/usr/local/share/bootloader-universe/_My_Operating_System_/`.
```

> You need to set SELinux to permissive when testing.

Reviewer 2: ```suggestion
Also you need to set SELinux to permissive.
```

> Create the following directory structure on the server according to your needs and supported operating systems:

Reviewer 3: ```suggestion
.Procedure
. Create the following directory structure on your {ProjectServer} according to your needs and supported operating systems:
```

[options="nowrap" subs="+quotes"]
----

> [root@or ~]# install -o foreman-proxy -g foreman-proxy -d /usr/local/share/bootloader-universe/{almalinux,rocky_linux,ubuntu}/

Reviewer 3: ```suggestion
# install -o foreman-proxy -g foreman-proxy -d /usr/local/share/bootloader-universe/{almalinux,rocky_linux,ubuntu}/
```

I'd also go with `_My_Operating_Systems_` or mark this as example.

Where does "almalinux" come from? Is alma-linux or "rocky-linux" als correct? Does this have to match the OS name from facter?

Author: As all commands will be executed on the same machine (=Foreman server) it's OK to strip the prompt.

> Where does "almalinux" come from? Is alma-linux or "rocky-linux" als correct? Does this have to match the OS name from facter?

https://github.com/theforeman/foreman-documentation/pull/2145/files#diff-2dae64f772bb99ae82fd42450722362d1d035b858664cdbc26ce5d0c9df881c5R71-R78

----
The subdirectories of operating systems you want to support must correspond to the names of the operating systems in lower case.

Example how to find the correct directory name:
[options="nowrap" subs="+quotes"]
----
[root@orcharhino ~]# hammer os list | grep "Rocky Linux" | tail -n1 | cut -d ' ' -f1 | xargs hammer os info --id | sed -nE "s/^Name:\s+(.*)$/\L\1/p"
rocky_linux
----

Extract shim and GRUB2 binaries from corresponding distribution packages and copy them into their subdirectories.
Example how to extract shim and GRUB2 binaries (Rocky Linux):
[options="nowrap" subs="+quotes"]
----

> [root@or /tmp]# wget http://dl.rockylinux.org/pub/rocky/9/BaseOS/x86_64/os/Packages/s/shim-x64-15.6-1.el9.rocky.0.2.x86_64.rpm

Reviewer 3: Please drop the prefix;only "#" is required.
